src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java

```diff
@@ -126,20 +143,107 @@ public String getPublishReportSnippet(ArrayList<String> paths){
         return snippet.toString();
     }
 
+    private String getSteps(ArrayList<Step> steps) throws InvocationTargetException, NoSuchMethodException, InstantiationException, ConfiguratorException, IllegalAccessException, NoSuchFieldException {
+        StringBuilder snippet = new StringBuilder();
+
+        snippet.append("\tsteps {\n");
+        for(Step step: steps)
+            snippet.append("\t\t").append(stepConfigurator(step));
+        snippet.append("\t}\n");
+
+        return snippet.toString();
+    }
+
+    public boolean isUnix(){
+        return this.launcher.isUnix();
+    }
+
+    private String completeShellScriptPath(String scriptPath){
+        if (isUnix()) {
+            return scriptPath + ".sh";
+        } else {
+            return scriptPath + ".bat";
+        }
+    }
+
+    private String stepConfigurator(Step step)
+            throws NoSuchMethodException, IllegalAccessException, InvocationTargetException,
+            InstantiationException, ConfiguratorException, NoSuchFieldException {
+        String snippet;
+        Descriptor stepDescriptor = StepDescriptor.byFunctionName(step.getStepName());
+
+        if(step.getStepName().equals("sh")){
+            if(step.getDefaultParameter() != null) {
+                step.setDefaultParameter(completeShellScriptPath(step.getDefaultParameter()));
+            }
+            else{
+                step.getParameters().put("script", completeShellScriptPath(step.getParameters().get("script").toString()));
+            }
+        }
+
+        if(stepDescriptor == null)
+            throw new RuntimeException(new IllegalStateException("No step exist with the name of " + step.getStepName()));
+
+        Class clazz = stepDescriptor.clazz;
+
+        Object stepObject = null;
```

Comment:
I think the order in which things happen here doesn't seem clear to me. I would think something like this would be more clear and do a better job of failing fast:
```
Object stepObject = null;

Descriptor stepDescriptor = ...
if(stepDescriptor == null)
    throw new ....
Class clazz = stepDescriptor.clazz;

if(step.getStepName().equals("sh")){
    ...
}
if(step.getDefaultParameter() != null){
    ...
}
else{
    ...
}
```

Reply:
[Early return](https://softwareengineering.stackexchange.com/questions/18454/should-i-return-from-a-function-early-or-use-an-if-statement) and early input validation is definitely the way to write code. It is much more readable.

Reply:
You can also do [precondition checks](https://github.com/twitter/commons/blob/master/src/java/com/twitter/common/styleguide.md#preconditions) which gets the bad inputs out of the way so you don't have to deal with then with nested if statements.